Add FasterTokenizer model in experiment

Steffy-zxf opened the pull request:
### PR types

 New features

### PR changes

APIs 

### Description

1. add FasterTokenizer model in experiment
2. add demo with FasterTokenizer usage in experiment

Commit:
1. add FasterTokenizer model in experiment
2. add demo with FasterTokenizer usage in experiment

Commit:
add cpp deploy demo

Commit:
add test time script

Commit:
add example

Commit:
mv python_deploy location

ZeyuChen commented on examples/experimental/faster_ernie/cpp_deploy/compile.sh:
```diff
@@ -0,0 +1,41 @@
+rm -rf *
+
+# same with the demo.cc
+DEMO_NAME=demo
```
不要叫DEMO，这不是DEMO

Steffy-zxf replied:
修改为text_cls_infer

ZeyuChen commented on examples/experimental/faster_ernie/cpp_deploy/demo.cc:
```diff
@@ -0,0 +1,64 @@
+
```
不要这些莫名其妙的空行

ZeyuChen replied:
文件名不要定义为demo，改为infer.cc

ZeyuChen replied:
或者是ernie_infer。同时后面应该还得区分下句子分类还是序列标注任务

Steffy-zxf replied:
done， 修改为text_cls_infer

ZeyuChen replied:
seq_cls_infer/token_cls_infer可能可以跟类名保持更好一致

ZeyuChen commented on examples/experimental/faster_ernie/cpp_deploy/demo.cc:
```diff
@@ -0,0 +1,64 @@
+      "办理入住手续，节省时间。"};
+
+  std::vector<float> probs;
+  Run(predictor.get(), &data, &probs);
```
需要要给出print的结果

ZeyuChen replied:
输出应该是const引用，保持data输入

ZeyuChen replied:
如果这个demo就是分类，那就写清楚分类的，和序列标注的分开

ZeyuChen replied:
seq_cls_infer/token_cls_infer可能可以跟类名保持更好一致

Steffy-zxf replied:
Done，已修改为seq_cls_infer

ZeyuChen commented on examples/experimental/faster_ernie/cpp_deploy/demo.cc:
```diff
@@ -0,0 +1,64 @@
+}
+
+void Run(Predictor* predictor,
+         std::vector<std::string>* input_data,
```
输入应该是用const引用，输出才是指针

Steffy-zxf replied:
修改为
```
void Run(Predictor* predictor,
         const std::vector<std::string>& input_data,
         std::vector<float>* logits,
         std::vector<int64_t>* predictions) 
```

ZeyuChen commented on examples/experimental/faster_ernie/export_model.py:
```diff
@@ -0,0 +1,54 @@
+
+import paddle
+import paddlenlp
+from paddlenlp.experimental import FastSequenceClassificationModel
```
Faster，我们整个技术代号统一使用Faster

Steffy-zxf replied:
Done

ZeyuChen commented on examples/experimental/faster_ernie/predict.py:
```diff
@@ -0,0 +1,91 @@
+import paddle
+import paddle.nn.functional as F
+from paddlenlp.ops import to_string_tensor
+from paddlenlp.experimental import FastSequenceClassificationModel
```
Faster

Steffy-zxf replied:
Done

ZeyuChen commented on examples/experimental/faster_ernie/predict.py:
```diff
@@ -0,0 +1,91 @@
+
+    Args:
+        model (obj:`paddle.nn.Layer`): A model to classify texts.
+        data (obj:`List(Example)`): The processed data whose each element is a Example (numedtuple) object.
```
这个注释对的上么，还有Examples这东西么

Steffy-zxf replied:
已删除

ZeyuChen commented on examples/experimental/faster_ernie/export_model.py:
```diff
@@ -0,0 +1,54 @@
+    model.eval()
+
+    # Convert to static graph with specific input description
+    model = paddle.jit.to_static(
```
这行to_static封装到`FasterErnieForSequenceClassification`

ZeyuChen replied:
我感觉动转静应该自动导出到softmax之后，而不是事后还需要自己再softmax

Steffy-zxf replied:
Done

ZeyuChen commented on examples/text_classification/pretrained_models/train.py:
```diff
@@ -117,15 +117,14 @@ def do_train():
 
     # If you wanna use bert/roberta/electra pretrained model,
     # tokenizer = ppnlp.transformers.BertTokenizer.from_pretrained('bert-base-chinese')
     # tokenizer = ppnlp.transformers.RobertaTokenizer.from_pretrained('roberta-wwm-ext')
```
去掉这些注释吧

Steffy-zxf replied:
已删除

ZeyuChen commented on examples/text_classification/pretrained_models/train.py:
```diff
@@ -117,15 +117,14 @@ def do_train():
     # model = ppnlp.transformers.RobertaForSequenceClassification.from_pretrained('roberta-wwm-ext', num_classes=2)
     # model = ppnlp.transformers.ElectraForSequenceClassification.from_pretrained('chinese-electra-small', num_classes=2)
```
去掉这些注释吧

Steffy-zxf replied:
已删除

ZeyuChen commented on paddlenlp/ops/strings.py:
```diff
@@ -12,10 +12,13 @@
+import paddle
 import paddle.fluid.core as core
 
+__all__ = ['to_string_tensor', 'to_vocab_tensor']
```
不要暴露vocab_tensor这个接口

ZeyuChen replied:
只暴露vocab即可

ZeyuChen replied:
to_string_tensor -> to_tensor
同时移入experimental中

Steffy-zxf replied:
Done

ZeyuChen commented on paddlenlp/ops/__init__.py:
```diff
@@ -25,3 +28,5 @@
 paddle.nn.TransformerEncoderLayer._ori_forward = paddle.nn.TransformerEncoderLayer.forward
 paddle.nn.TransformerEncoder._ori_forward = paddle.nn.TransformerEncoder.forward
+
+Strings = core.VarDesc.VarType.STRINGS
```
这行代码有什么用？

Steffy-zxf replied:
为了封装STRINGS这个类型，为了导出模型用的。已删除Strings，在paddlenlp库内部调用`core.VarDesc.VarType.STRINGS`，不对外暴露。

ZeyuChen commented on paddlenlp/experimental/faster_tokenizer.py:
```diff
@@ -0,0 +1,74 @@
+class FasterTokenizer(nn.Layer):
+    def __init__(self, vocab, do_lower_case=False, is_split_into_words=False):
+        super(FasterTokenizer, self).__init__()
+        vocab_tensor = to_vocab_tensor(vocab, "vocab")
```
to_vocab_buffer

Steffy-zxf replied:
Done

ZeyuChen commented on paddlenlp/experimental/faster_tokenizer.py:
```diff
@@ -0,0 +1,74 @@
+    def __init__(self, vocab, do_lower_case=False, is_split_into_words=False):
+        super(FasterTokenizer, self).__init__()
+        vocab_tensor = to_vocab_tensor(vocab, "vocab")
+        self.register_buffer("vocab", vocab_tensor, persistable=True)
```
vocab_tensor -> vocab_buffer

Steffy-zxf replied:
Done

ZeyuChen commented on examples/experimental/faster_tokenizer/demo.py:
```diff
@@ -0,0 +1,73 @@
+text = '小说是文学的一种样式，一般描写人物故事，塑造多种多样的人物形象，但亦有例外。它是拥有不完整布局、发展及主题的文学作品。而对话是不是具有鲜明的个性，每个人物说的没有独特的语言风格，是衡量小说水准的一个重要标准。与其他文学样式相比，小说的容量较大，它可以细致的展现人物性格和命运，可以表现错综复杂的矛盾冲突，同时还可以描述人物所处的社会生活环境。小说一词，最早见于《庄子·外物》：“饰小说以干县令，其于大达亦远矣。”这里所说的小说，是指琐碎的言谈、小的道理，与现时所说的小说相差甚远。文学中，小说通常指长篇小说、中篇、短篇小说和诗的形式。小说是文学的一种样式，一般描写人物故事，塑造多种多样的人物形象，但亦有例外。它是拥有不完整布局、发展及主题的文学作品。而对话是不是具有鲜明的个性，每个人物说的没有独特的语言风格，是衡量小说水准的一个重要标准。与其他文学样式相比，小说的容量较大，它可 ...
+data = [text[:max_seq_length]] * 100
+
+text_tensor = to_string_tensor(data)
```
这个变量做什么的？没看见使用

Steffy-zxf replied:
已删除

ZeyuChen commented on examples/experimental/faster_ernie/train.py:
```diff
@@ -0,0 +1,174 @@
+
+    train_ds, dev_ds = load_dataset("chnsenticorp", splits=["train", "dev"])
+
+    model = FastSequenceClassificationModel.from_pretrained(
```
FasterModelForSequenceClassification

ZeyuChen replied:
FasterErnieForSequenceClassification

Steffy-zxf replied:
已修改为FasterModelForSequenceClassification

Commit:
optimize cpp demo

Commit:
add faster model

Commit:
add faster tokenizer export model

ZeyuChen commented on examples/experimental/faster_ernie/python_deploy.py:
```diff
@@ -0,0 +1,147 @@
+        model_file = os.path.join(model_dir, "inference.pdmodel")
+        params_file = os.path.join(model_dir, "inference.pdiparams")
+        if not os.path.exists(model_file):
+            raise ValueError("not find model file path {}".format(model_file))
```
Error log与注释一样风格，保持首字母大写的标准英文句式

ZeyuChen replied:
然后应该是XXXX file not found比较准确。

ZeyuChen commented on examples/experimental/faster_ernie/python_deploy.py:
```diff
@@ -0,0 +1,147 @@
+        logits = self.output_handle.copy_to_cpu()
+
+        # probs = softmax(logits, axis=1)
+        # idx = np.argmax(probs, axis=1)
```
需要评估下这种分类，是否把probs和argmax的结果同时导出比较好

Steffy-zxf replied:
同时导出logits，predictins(index)

ZeyuChen commented on examples/experimental/faster_ernie/train.py:
```diff
@@ -0,0 +1,174 @@
+from paddlenlp.datasets import load_dataset
+from paddlenlp.transformers import LinearDecayWithWarmup
+from paddlenlp.ops import to_string_tensor
+from paddlenlp.experimental import FastSequenceClassificationModel
```
FasterErnieForSequenceClassification

ZeyuChen replied:
or FasterModelForSequenceClassification
不要与正式版本的命名风格存在太大差一

Steffy-zxf replied:
FasterModelForSequenceClassification

ZeyuChen commented on paddlenlp/experimental/model.py:
```diff
@@ -0,0 +1,164 @@
+        return logits
+
+
+class FastSequenceClassificationModel(object):
```
FasterModelForSequenceClassification

Steffy-zxf replied:
Done

ZeyuChen commented on paddlenlp/ops/strings.py:
```diff
@@ -12,10 +12,13 @@
+import paddle
 import paddle.fluid.core as core
 
+__all__ = ['to_string_tensor', 'to_vocab_tensor']
```
整体挪到paddlenlp.experimental中去

Steffy-zxf replied:
Done

ZeyuChen commented on paddlenlp/experimental/model.py:
```diff
@@ -0,0 +1,164 @@
+            raise ValueError("Unknown name %s. Now %s surports  %s" %
+                             (pretrained_model_name_or_path, cls.__name__,
+                              list(name_model.keys())))
+
```
基于这个类新增to_static接口，屏蔽STRINGS类型对外暴露

Steffy-zxf replied:
Done

ZeyuChen commented on paddlenlp/experimental/model.py:
```diff
@@ -0,0 +1,164 @@
+        else:
+            raise ValueError("Unknown name %s. Now %s surports  %s" %
+                             (pretrained_model_name_or_path, cls.__name__,
+                              list(name_model.keys())))
```
基于这个类新增to_static接口，屏蔽STRINGS类型对外暴露

Steffy-zxf replied:
Done

ZeyuChen commented on paddlenlp/ops/strings.py:
```diff
@@ -12,10 +12,13 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import paddle
```
移入paddlenlp/experimental/中

Steffy-zxf replied:
Done

ZeyuChen reviewed (commented):
to_string_tensor -> to_tensor, 移入experimental中
动转静需要内置到FasterModelForXXXX类中，在上层动转静接口中屏蔽STRINGS对象暴露
动转静导出建议同时导出probs和argmax后结果，可以使推理结果更加便捷。
不要让用户额外撰写softmax的算子实现

Commit:
1. rename to_string_tensor to to_tensor
2. rename to_vocab_tensor to to_vocab_buff
3. add to_static() api to FasterModel

Commit:
1. add text cls example with FasterModelForSequenceClassification
2. remove some redudant code

ZeyuChen commented on examples/experimental/faster_ernie/text_cls/cpp_deploy/CMakeLists.txt:
```diff
@@ -0,0 +1,221 @@
+    set(CUDA_LIB "/usr/local/cuda/lib64/" CACHE STRING "CUDA Library")
+  else()
+    if(CUDA_LIB STREQUAL "")
+      set(CUDA_LIB "C:\\Program\ Files\\NVIDIA GPU Computing Toolkit\\CUDA\\v8.0\\lib\\x64")
```
这个命令写死路径可能不一定正确，回头得windows测试验证下

ZeyuChen commented on examples/experimental/faster_ernie/text_cls/train.py:
```diff
@@ -0,0 +1,162 @@
+        losses.append(loss.numpy())
+        correct = metric.compute(logits, labels)
+        metric.update(correct)
+        accu = metric.accumulate()
```
此处的accumulte应该在循环外还是在循环内？

ZeyuChen commented on examples/experimental/faster_ernie/text_cls/train.py:
```diff
@@ -0,0 +1,162 @@
+
+
+def create_dataloader(dataset, mode='train', batch_size=1):
+
```
去掉无用空行

ZeyuChen commented on examples/experimental/faster_tokenizer/demo.py:
```diff
@@ -0,0 +1,92 @@
+text = '小说是文学的一种样式，一般描写人物故事，塑造多种多样的人物形象，但亦有例外。它是拥有不完整布局、发展及主题的文学作品。而对话是不是具有鲜明的个性，每个人物说的没有独特的语言风格，是衡量小说水准的一个重要标准。与其他文学样式相比，小说的容量较大，它可以细致的展现人物性格和命运，可以表现错综复杂的矛盾冲突，同时还可以描述人物所处的社会生活环境。小说一词，最早见于《庄子·外物》：“饰小说以干县令，其于大达亦远矣。”这里所说的小说，是指琐碎的言谈、小的道理，与现时所说的小说相差甚远。文学中，小说通常指长篇小说、中篇、短篇小说和诗的形式。小说是文学的一种样式，一般描写人物故事，塑造多种多样的人物形象，但亦有例外。它是拥有不完整布局、发展及主题的文学作品。而对话是不是具有鲜明的个性，每个人物说的没有独特的语言风格，是衡量小说水准的一个重要标准。与其他文学样式相比，小说的容量较大，它可 ...
+data = [text[:max_seq_length]] * 100
+
+pp_tokenizer = FasterTokenizer(vocab, do_lower_case=False)
```
此处接口是否需要与XXXTokenizer.from_pretrained的API体验打平？以及是否需要

Commit:
update cpp deploy

Commit:
1. rename text_cls with seq_cls
2. suuport from_pretrained() with given a local directory

Commit:
add experimental.model_utils.py

Commit:
update seq_cls cpp demo, drop to print the label string

Commit:
add FasterTokenizer.from_pretrained() api

Commit:
use FasterTokenizer.from_pretrained() api

ZeyuChen reviewed (approved):
LGTM

ZeyuChen committed:
Merge branch 'develop' into exp-faster-tokenizer
